vnext/Desktop/Modules/WebSocketModule.cpp

```diff
@@ -13,123 +14,175 @@
+          }
+        }
+
+        weak_ptr weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 3), jsArgAsString(args, 0));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->Connect(protocols, options);
+        }
+      }),
+    Method(
+      "close",
+      [this](dynamic args) // [int64_t code, string reason,] int64_t id
+      {
+        // See react-native\Libraries\WebSocket\WebSocket.js:_close
+        if (args.size() == 3) // WebSocketModule.close(statusCode, closeReason, this._socketId);
+        {
+          weak_ptr weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 2));
+          if (auto sharedWs = weakWs.lock())
+          {
+            sharedWs->Close(static_cast<IWebSocketResource::CloseCode>(jsArgAsInt(args, 0)), jsArgAsString(args, 1));
+          }
+        }
+        else if (args.size() == 1) // WebSocketModule.close(this._socketId);
+        {
+          weak_ptr weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 0));
+          if (auto sharedWs = weakWs.lock())
+          {
+            sharedWs->Close(IWebSocketResource::CloseCode::Normal, {});
+          }
+        }
+        else
+        {
+          auto errorObj = dynamic::object("id", -1)("message", "Incorrect number of parameters");
+          this->SendEvent("websocketFailed", std::move(errorObj));
+        }
+      }),
+    Method(
+      "send",
+      [this](dynamic args) // const string& message, int64_t id
+      {
+        weak_ptr weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 1));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->Send(jsArgAsString(args, 0));
+        }
+      }),
+    Method(
+      "sendBinary",
+      [this](dynamic args) // const string& base64String, int64_t id
+      {
+        weak_ptr weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 1));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->SendBinary(jsArgAsString(args, 0));
+        }
+      }),
+    Method(
+      "ping",
+      [this](dynamic args) // int64_t id
+      {
+        weak_ptr weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 0));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->Ping();
+        }
+      })
+  };
 } // getMethods
 
 #pragma region private members
 
-void WebSocketModule::SendEvent(string &&eventName, dynamic &&args) {
+void WebSocketModule::SendEvent(string&& eventName, dynamic&& args)
+{
   auto instance = this->getInstance().lock();
   if (instance)
+  {
     instance->callJSFunction("RCTDeviceEventEmitter", "emit", dynamic::array(std::move(eventName), std::move(args)));
+  }
 }
 
-IWebSocketResource *WebSocketModule::GetOrCreateWebSocket(int64_t id, string &&url) {
-  IWebSocketResource *ptr = nullptr;
-  if (m_webSockets.find(id) == m_webSockets.end()) {
+std::shared_ptr<IWebSocketResource> WebSocketModule::GetOrCreateWebSocket(int64_t id, string&& url)
+{
+  auto itr = m_webSockets.find(id);
```

Comment:
>auto [](start = 2, length = 4)

can this be const& ?

Reply:
I disagree with this advice. Const refs have some special lifetime extending rules, but if this was a normal ref we'd have lifetime issues and that distinction is often too subtle.

Reply:
@NickGerleman Can you explain (or link to some explanation) of the lifetime rules you're describing?

Reply:
https://en.cppreference.com/w/cpp/language/lifetime

> All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created, and if multiple temporary objects were created, they are destroyed in the order opposite to the order of creation. This is true even if that evaluation ends in throwing an exception.
>There are two exceptions from that:
The lifetime of a temporary object may be extended by binding to a const lvalue reference or to an rvalue reference (since C++11), see reference initialization for details.

I've seen this trip up some folks who are relatively experienced. Keeping the iterator by value avoids the ambiguity of coercing the returned value into a reference.